core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -172,6 +174,130 @@ void *CudaSpace::allocate(const char *arg_label, const size_t arg_alloc_size,
+  error_code = cudaSuccess;
+
+  // Interpret env var
+  char *env_string = getenv("EXPERIMENTAL_CUDA_ASYNC_MEMPOOL_SIZE");
+  if (!env_string) return true;  // Nothing to set, initialization is complete
+
+  // Env var can be a decimal e.g. 2.4g
+  double factor = 1;  // bytes
+
+  // Interpret the string
+  std::string mempool_size_string(env_string);
+  char last_letter = mempool_size_string.back();
+
+  // We can deal with the last letter being units.
+  if (std::isalpha(last_letter)) {
+    mempool_size_string.pop_back();
+    char last_letter_cap = std::toupper(last_letter);
+    switch (last_letter_cap) {
+      case 'K': factor = 1024; break;
+      case 'M': factor = 1024 * 1024; break;
+      case 'G': factor = 1024 * 1024 * 1024; break;
+      default:
+        std::cerr << "Unknown units for requested size " << last_letter << "\n";
+        return false;
+        break;
+    }
+  }
+
+  // At this point we removed the last letter (or there was not a last letter)
+  // convert what is left to a number.
+  double requested_size = 0;
+
+  // Handle exception in case the string is unconvertible
+  try {
+    requested_size = std::stod(mempool_size_string);
+  } catch (...) {
+    std::cerr << "Unable to convert " << mempool_size_string
+              << " to a number\n";
+    return false;
+  }
+
+  // Check for negative size memory requests (zero is allowed)
+  if (requested_size < 0) {
+    std::cerr << "Negative amount of memory requested in allocation\n";
+    return false;
+  }
+
+  // Multiply in units
+  requested_size *= factor;
+
+  // Check we are not asking for memory that is greater than what size_t can
+  // hold. Since the requested size can be the larger I convert the maximum of
+  // size_t to a double and compare those
+  double max_size_t = static_cast<double>(std::numeric_limits<size_t>::max());
+  if (requested_size > max_size_t) {
+    std::cerr << "Requested amount of memory " << requested_size << " exceeds "
+              << "maximum allocatable size " << max_size_t << "\n";
+    return false;
+  }
+
+  // At this point requested_size should be appropriate
+  // neither too big nor negative - even with the ceiling it should
+  // not be bigger than max_size_t. Safe to cast
+  size_t n_bytes = static_cast<size_t>(std::ceil(requested_size));
+
+  // We set up the default memory pool
+  // This requires us to a) grab the default memory pool for the device
+  // b) set a property to retain our desired amount of memory on a free
+  // c) allocate our desired poolsize amount plus a little more
+  // d) free the memory, ensuring the pool retains what we asked for
+  // All the API calls need to return cudaSuccess or we fail
+
+  // First get the default memory pool
+  error_code = cudaSuccess;
+  cudaMemPool_t default_mempool;  // The memory pool handle
+
+  // Cuda API call to get the memory pool for device with device_id
+  error_code = cudaDeviceGetDefaultMemPool(&default_mempool, device_id);
```

Comment:
As `Cuda` allows creating memory pool, and actually allows much more attributes to be set, I would propose to get this code a bit more "structured". Something like:
```c++
class CudaMemoryPool {
    cudaMemPool_t handle;

    static auto get_default(const Cuda& exec) {
        CudaMemoryPool pool;
        KOKKOS_IMPL_CUDA_SAFE_CALL(cudaDeviceGetDefaultMemPool(&pool, exec.device_id));
        return pool;
    }

    template <cudaMemPoolAttr Attribute, typename ValueType>
    void set(const ValueType& value) const {
        KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMemPoolSetAttribute(handle, Attribute, &value));
    }

    void initialize(const Cuda& exec, const size_t n_bytes) const {
        void *dummy_ptr = nullptr;
        KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMallocAsync(&dummy_ptr, n_bytes, exec.stream));
        KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFreeAsync(dummy_ptr, exec.stream));
    }
};

std::unordered_set<CudaMemoryPool> memory_pools;
```